`createNvCapDevice()`: may create `/dev/nvidia-caps/nvidia-cap<xxx>` with bad major number

In `createNvCapDevice()` we [read](https://github.com/NVIDIA/k8s-dra-driver-gpu/blob/085c1797e7ce52425a4b61d200c50dfbb5ad56c4/cmd/compute-domain-kubelet-plugin/nvlib.go#L348) the host's `/proc/devices` file in order to obtain the `major` and `minor` device number of the character device called `nvidia-caps`.

Parsing is done by https://github.com/NVIDIA/k8s-dra-driver-gpu/blob/085c1797e7ce52425a4b61d200c50dfbb5ad56c4/cmd/compute-domain-kubelet-plugin/nvlib.go#L228

On an example host, `/proc/devices` may look like:
```
$ cat /proc/devices
Character devices:
  1 mem
<snip>
234 nvidia-caps
235 nvidia-caps-imex-channels
236 nvme-generic
<snip>
504 NVDA2004:00
<snip>

Block devices:
  7 loop
<snip>
```

Current parser code may leave the parsing loop early when a device is present that contains a colon in its name (as in case of `NVDA2004:00` in the example output above):

https://github.com/NVIDIA/k8s-dra-driver-gpu/blob/085c1797e7ce52425a4b61d200c50dfbb5ad56c4/cmd/compute-domain-kubelet-plugin/nvlib.go#L247

In that case, we [return](https://github.com/NVIDIA/k8s-dra-driver-gpu/blob/085c1797e7ce52425a4b61d200c50dfbb5ad56c4/cmd/compute-domain-kubelet-plugin/nvlib.go#L269C2-L269C26)

```
-1, scanner.Err()
```
whereas `-1` is what the caller reads a major number, and `scanner.Err()` is not an error (but `nil`) because scanning (reading the file) did not result in any I/O error or other error.

About `scanner.Err()`:

> Err returns the first non-EOF error that was encountered by the [Scanner](https://pkg.go.dev/bufio#Scanner).

(https://pkg.go.dev/bufio#Scanner.Err)

Consequently, this error handler does _not_ hit in:
```
	major, err := l.getDeviceMajor(nvidiaCapsDeviceName)
	if err != nil {
		return nil, fmt.Errorf("error getting device major: %w", err)
	}
```
and we continue processing `-1` as major number.

A signed integer is often represented as "two's complement" in memory. In that case the bitwise representation of -1 is all ones:

> in two’s complement if all the bits of an integer are 1, then the integer is equal to −1, regardless of how many bits there are

In `createNvCapDevice()` we then [create](https://github.com/NVIDIA/k8s-dra-driver-gpu/blob/085c1797e7ce52425a4b61d200c50dfbb5ad56c4/cmd/compute-domain-kubelet-plugin/nvlib.go#L356) a dev node with a major number of `-1`.  As explained, all bits are set as `1`, i.e. all bytes have value `F` in hex notation.

That is why on the created dev node, `stat` reports the device major as `fff`:
```
# stat /dev/nvidia-caps/nvidia-cap4323 
...
Device: 5h/5d	Inode: 2678        Links: 1     Device type: fff,10e3
...
```
`fff` is 4095 in decimal, so `ls` reports:
```
# ls -la /dev/nvidia-caps
...
cr--------  1 root root 4095, 4323 Mar 25 20:34 nvidia-cap4323
```

As this value is wrong (not the expected / correct device major number), IMEX daemon startup after all fails with a permission error:

> [Mar 25 2025 17:06:27] [ERROR] [tid 38] Failed to acquire required permissions to access driver. If you are running nvidia-imex from a non-root context, make sure required NVIDIA GPU driver capability file permissions are assigned as referenced in user guide.

(also see related issue https://github.com/NVIDIA/k8s-dra-driver-gpu/issues/295#issuecomment-2751970858).

This is a beautiful text-parsing and error handling bug. As of the nature of per-host individual contents of `/proc/devices`, the problem may be rare (does not appear on every host).

--- a/cmd/compute-domain-kubelet-plugin/nvlib.go
+++ b/cmd/compute-domain-kubelet-plugin/nvlib.go
@@ -253,15 +253,20 @@ func (l deviceLib) getDeviceMajor(name string) (int, error) {
 		return -1, fmt.Errorf("error parsing '%s': unexpected regex match: %v", procDevicesPath, matches)
 	}
 
-	// Convert capture group content to integer. Error should be unreachable
-	// (because capture group has matched [0-9]+ which should always convert to
-	// integer).
-	major, err := strconv.Atoi(matches[1])
+	// Convert capture group content to integer. Perform upper bound check:
+	// value must fit into 32-bit integer (it's then also guaranteed to fit into
+	// a 32-bit unsigned integer, which is the type that must be passed to
+	// unix.Mkdev()).
+	major, err := strconv.ParseInt(matches[1], 10, 32)
 	if err != nil {
 		return -1, fmt.Errorf("int conversion failed for '%v': %w", matches[1], err)
 	}
 
-	return major, nil
+	// ParseInt() always returns an integer of explicit type `int64`. We have
+	// performed an upper bound check so it's safe to convert this to `int`
+	// (which is documented as "int is a signed integer type that is at least 32
+	// bits in size", so in theory it could be smaller than int64).
+	return int(major), nil
 }
 
 func (l deviceLib) parseNVCapDeviceInfo(nvcapsFilePath string) (*nvcapDeviceInfo, error) {

--- a/cmd/compute-domain-kubelet-plugin/nvlib.go
+++ b/cmd/compute-domain-kubelet-plugin/nvlib.go
@@ -22,6 +22,7 @@ import (
 	"os"
 	"os/exec"
 	"path/filepath"
+	"regexp"
 	"strconv"
 	"strings"
 
@@ -225,48 +226,42 @@ func (l deviceLib) getImexChannelCount() (int, error) {
 	return 2048, nil
 }
 
+// getDeviceMajor searches for one "<integer> <name>" occurrence in the
+// "Character devices" section of the /proc/devices file, and returns the
+// integer.
 func (l deviceLib) getDeviceMajor(name string) (int, error) {
-	file, err := os.Open(procDevicesPath)
+
+	re := regexp.MustCompile(
+		// The `(?s)` flag makes `.` match newlines. The greedy modifier in
+		// `.*?` ensures to pick the first match after "Character devices".
+		// Extract the number as capture group (the first and only group).
+		"(?s)Character devices:.*?" +
+			"([0-9]+) " + regexp.QuoteMeta(name) +
+			".*Block devices:",
+	)
+
+	data, err := os.ReadFile(procDevicesPath)
 	if err != nil {
-		return -1, err
-	}
-	defer file.Close()
-
-	scanner := bufio.NewScanner(file)
-	foundCharDevices := false
-
-	for scanner.Scan() {
-		line := scanner.Text()
-
-		// Ignore empty lines
-		if line == "" {
-			continue
-		}
-
-		// Check for any line with text followed by a colon (header)
-		if strings.Contains(line, ":") {
-			// Stop if we've already found the character devices section and reached another section
-			if foundCharDevices {
-				break
-			}
-			// Check if we entered the character devices section
-			if strings.HasSuffix(line, ":") && strings.HasPrefix(line, "Character") {
-				foundCharDevices = true
-			}
-			// Continue to the next line, regardless
-			continue
-		}
-
-		// If we've passed the character devices section, check for nvidiaCapsImexChannelsDeviceName
-		if foundCharDevices {
-			parts := strings.Fields(line)
-			if len(parts) == 2 && parts[1] == name {
-				return strconv.Atoi(parts[0])
-			}
-		}
+		return -1, fmt.Errorf("error reading '%s': %w", procDevicesPath, err)
 	}
 
-	return -1, scanner.Err()
+	// Expect precisely one match: first element is the total match, second
+	// element corresponds to first capture group within that match (i.e., the
+	// number of interest).
+	matches := re.FindStringSubmatch(string(data))
+	if len(matches) != 2 {
+		return -1, fmt.Errorf("error parsing '%s': unexpected regex match: %v", procDevicesPath, matches)
+	}
+
+	// Convert capture group content to integer. Error should be unreachable
+	// (because capture group has matched [0-9]+ which should always convert to
+	// integer).
+	major, err := strconv.Atoi(matches[1])
+	if err != nil {
+		return -1, fmt.Errorf("int conversion failed for '%v': %w", matches[1], err)
+	}
+
+	return major, nil
 }
 
 func (l deviceLib) parseNVCapDeviceInfo(nvcapsFilePath string) (*nvcapDeviceInfo, error) {

